source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +119,281 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+
+        if (tx.outputs.length == 0)
+            return "GenesisBlock: No output(s) in the transaction";
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, const ref output; tx.outputs)
+        {
+            // disallow negative amounts
+            if (!output.value.isValid())
+                return "GenesisBlock: Output(s) overflow or underflow"
+                    ~ "in the transaction";
+
+            // disallow 0 amount
+            if (output.value == Amount(0))
+                return "GenesisBlock: Value of output is 0"
+                    ~ "in the transaction";
+
+            const UTXOSetValue utxo_value = {
+                unlock_height: 0,
+                type: tx.type,
+                output: output
+            };
+            utxo_set[UTXOSetValue.getHash(tx_hash, idx)] = utxo_value;
+        }
+    }
+
+    Hash[] merkle_tree;
+    if (block.header.merkle_root !=
+        Block.buildMerkleTree(block.txs, merkle_tree))
+        return "GenesisBlock: Merkle root does not match header's";
+
+    // If there are no enrollments, return them here early.
+    if (block.header.enrollments.length == 0)
+        return null;
+
+    if (!isStrictlyMonotonic!"a.utxo_key < b.utxo_key"
+        (block.header.enrollments))
+        return "GenesisBlock: The enrollments should be arranged in "
+            ~ "ascending order by the utxo_key";
+
+    Set!Hash used_utxos;
+    bool findUTXO (Hash hash, size_t index, out UTXOSetValue value)
+        nothrow @safe
+    {
+        Hash utxo_hash;
+
+        if (index == size_t.max)
+            utxo_hash = hash;
+        else
+            utxo_hash = UTXOSetValue.getHash(hash, index);
+
+        if (utxo_hash in used_utxos)
+            return false;  // double-spend
+
+        if (auto ptr = utxo_hash in utxo_set)
+        {
+            value = *ptr;
+            used_utxos.put(utxo_hash);
+            return true;
+        }
+        return false;
+    }
+
+    foreach (const ref enrollment; block.header.enrollments)
+    {
+        if (auto fail_reason = VEn.isInvalidReason(enrollment, &findUTXO))
+            return fail_reason;
+    }
+
+    return null;
+}
+
+/// Genesis block validation fail test
+unittest
+{
+    import agora.common.Serializer;
+
+    Block block = GenesisBlock.serializeFull.deserializeFull!Block;
+    assert(block.isGenesisBlockValid());
+
+    // don't accept block height 0 from the network
+    assert(!block.isValid(0, Hash.init, null));
+
+    KeyPair keypair;
+
+    // height check
+    block.header.height = 1;
+    assert(!block.isGenesisBlockValid());
+
+    block.header.height = 0;
+    assert(block.isGenesisBlockValid());
+
+    // .prev_block check
+    block.header.prev_block = block.header.hashFull();
+    assert(!block.isGenesisBlockValid());
+
+    block.header.prev_block = Hash.init;
+    assert(block.isGenesisBlockValid());
+
+    Transaction tx =
+        UnitTestGenesisTransaction.serializeFull.deserializeFull!Transaction;
```

Comment:
```suggestion
    Transaction[] txs = GenesisBlock.txs.serializeFull.deserializeFull!(Transaction[]);
```

Reply:
With this change you can go back to making `UnitTestGenesisTransaction` private again.

Reply:
That's right.